libtransmission/peer-io.h

```diff
@@ -66,12 +66,46 @@ struct evbuffer_deleter
 
 using tr_evbuffer_ptr = std::unique_ptr<evbuffer, evbuffer_deleter>;
 
-class tr_peerIo
+class tr_peerIo final : public std::enable_shared_from_this<tr_peerIo>
 {
     using DH = tr_message_stream_encryption::DH;
     using Filter = tr_message_stream_encryption::Filter;
 
 public:
+    ~tr_peerIo();
+
+    // TODO: 8 constructor args is too many; maybe a builder object?
+    static std::shared_ptr<tr_peerIo> newOutgoing(
+        tr_session* session,
+        tr_bandwidth* parent,
+        struct tr_address const* addr,
+        tr_port port,
+        time_t current_time,
+        tr_sha1_digest_t const& torrent_hash,
+        bool is_seed,
+        bool utp);
+
+    static std::shared_ptr<tr_peerIo> newIncoming(
+        tr_session* session,
+        tr_bandwidth* parent,
+        struct tr_address const* addr,
+        tr_port port,
+        time_t current_time,
+        struct tr_peer_socket const socket);
+
+    // this is only public for testing purposes.
+    // production code should use newOutgoing() or newIncoming()
+    static std::shared_ptr<tr_peerIo> create(
```

Comment:
So maybe just use `friend` for Test class?

Reply:
moved in b4bb51df173f8444e2382a21db55d4904890fb74